Implement "dockerd --dns-xxx" based dns mode

codablock opened the pull request:
This PR implements "dockerd --dns" based DNS as an alternative to dnsmasq.

It can be enabled by settings dns_mode to docker_kubedns. dns_mode is a replacement for skip_dnsmasq and setup_dns.

As I understand it, dnsmasq was initially introduced to make hostNetwork PODs have proper cluster DNS configured. Maybe there were other reasons that I missed, but my feeling is that the solution got too complex for what it tries to solve.

An alternative is to use the docker daemon options --dns, --dns-search and --dns-opt. The docker_kubedns mode will add these to the docker service and skip all other DNS related roles/tasks. This means, that the hosts /etc/resolv.conf and dhclient conf is not touched at all.

Kubelet gets the kubedns/skydns IP as --cluster_dns, so all normal (non-hostNetwork) PODs will generate their own resolv.conf that has proper cluster DNS set up. PODs that have 'hostNetwork: true' set, will rely on docker providing a correct resolv.conf, which will contain proper cluster DNS settings as well (due to --dns-xxx).

The docker_kubedns mode also parses the hosts resolv.conf to figure out which nameservers and search domains are used. These are added as proper --dns-xxx args as well. These nameservers come after the kubedns entry, which makes them backup nameservers. This basically makes the early_dns logic unnecessary, as the hosts resolv.conf is untouched and only docker containers are affected (lookup still works, but takes longer) by the not running cluster DNS server.

The long term goal of this PR is to maybe completely remove dnsmasq and all the modifications to resolv.con and dhclient.conf and then leave only the docker_kubedns mode. This would simplify a lot of things and remove a big source of error.

The default dns_mode is set to dnsmasq_kubedns, which is the same as skip_dnsmasq=false and setup_dns=true.

I still need to update the documentation. Will do this after review.

bogdando commented:
> The docker_kubedns mode also parses the hosts resolv.conf to figure out which nameservers and search domains are used. These are added as proper --dns-xxx args as well. These nameservers come after the kubedns entry, which makes them backup nameservers. 

The dnsmasq ds should still be an option to leave with the docker_mode. There is a limitation for max nameservers in resolv.conf, and one is taken vy k8s cluster dns ip, either kubedns/skydns or dnsmasq ds.

So I suggest to allow users use both `docker_mode` and  leave `skip_dnsmasq: False` as is.

codablock commented:
As already asked on Slack: Are there really use cases for more then 3 nameservers in resolv.conf? Even if it's reduced to 2 (due to skydns entry), why would anyone need more then 2?

Supporting rkt later should not be a problem as it supports the same dns flags.

codablock commented:
I just figured out that host networks with --dns only work with docker >= 1.12. https://github.com/docker/docker/pull/22408 added a the needed bit to make it work.

The strange thing is, that kargo always installed 1.12 for me even though docker_version is set to 1.12. I'm on CentOS 7 but I'd expect the same is true for RedHat. CI already passed for coreos-calico-sep and centos7-flannel. ubuntu-weave-sep is currently running and I expect it to also pass as the logs already show that docker 1.12 is installed on the nodes.

So...there seems to be a bug that ignores the docker_version variable. What should we do about it? And also, which version should we actually install?

codablock commented:
Another thing: When https://github.com/kubernetes/kubernetes/pull/29378 gets merged (I'd expect it for k8s 1.6.) all this dns stuff in kargo may get obsolete as it would allow to specify "dnsPolicy: ClusterFirstAlways" for hostNetwork PODs.

bogdando commented on cluster.yml:
```diff
@@ -55,8 +55,8 @@
   roles:
-    - { role: dnsmasq, tags: dnsmasq }
-    - { role: kubernetes/preinstall, tags: resolvconf }
+    - { role: dnsmasq, when: "dns_mode == 'dnsmasq_kubedns'", tags: dnsmasq }
```
As we discussed, this should remain unconditional

codablock replied:
This was unconditional before and had the effect of installing dnsmasq even if skip_dnsmasq was true.

I now changed the variables to:
```yaml
# Can be dnsmasq_kubedns, kubedns or none
dns_mode: dnsmasq_kubedns

# Can be docker_dns, host_resolvconf or none
hostnet_dns_mode: host_resolvconf
```

This should give us the most flexibility.

bogdando commented on contrib/terraform/openstack/group_vars/all.yml:
```diff
@@ -105,21 +105,18 @@ kube_apiserver_insecure_port: 8080 # (http)
-# Kubernetes won't do this for you (yet).
 
-# Do not install additional dnsmasq
-skip_dnsmasq: false
```
As we discussed, `skip_dnsmasq` logic should remain as is

codablock replied:
Please see https://github.com/kubernetes-incubator/kargo/pull/799#discussion_r93601475

bogdando replied:
Sorry but a hoped to see both docker dns and dnsmasq_kubedns applicable, which is:
* hosts' DNS config is not changed
* Pods DNS config points to the dnsmasq ds cluster ip

bogdando replied:
what's why I proposed to not change `skip_dnsmasq` logic, it has nothing to the hosts/pods config.
Please keep this var and use `dns_mode` to only control if hosts' DNS config will be managed or docker one, and also to supersede `dns_setup`

codablock replied:
This possible with:
```yaml
# Can be dnsmasq_kubedns, kubedns or none
dns_mode: dnsmasq_kubedns

# Can be docker_dns, host_resolvconf or none
resolvconf_mode: docker_dns
```

I decided to remove the skip_dnsmasq and setup_dns flags, as I always got confused by them. skip_dnsmasq had to be 'false' to enable dnsmasq and setup_dns had to be 'true' to enable it. I thought this was not consistent and may confuse people about what is possible and what not.

I hope with the 2 new modes, it's clearer that the dns mode (dnsmasq+kubedns or kubedns) and resolv conf related stuff are actually 2 different things. This is something that caused a lot of confusion on my side as you probably already have noticed in our discussions ;)

bogdando replied:
ok, just please update the table with new vars then

bogdando commented on roles/dnsmasq/meta/main.yml:
```diff
@@ -2,5 +2,5 @@
   - role: download
     file: "{{ downloads.dnsmasq }}"
-    when: not skip_dnsmasq|default(false) and download_localhost|default(false)
+    when: dns_mode == 'dnsmasq_kubedns' and download_localhost|default(false)
```
ditto

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,41 @@
+
+- name: add upstream dns servers
+  set_fact:
+    docker_dns_servers: "{{ docker_dns_servers + upstream_dns_servers|default([]) }}"
```
please use union and uniq

codablock replied:
uniq is dangerous here as it would change the order of the entries

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,41 @@
+
+- name: add global searchdomains
+  set_fact:
+    docker_dns_search_domains: "{{ docker_dns_search_domains + searchdomains|default([]) }}"
```
please use union and uniq. And add assert that it's length <=6

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,41 @@
+    docker_dns_search_domains: "{{ docker_dns_search_domains + searchdomains|default([]) }}"
+
+- name: check system nameservers
+  shell: grep "^nameserver" /etc/resolv.conf | sed 's/nameserver\s*//'
```
please use `s/^namserver.....`

codablock replied:
done

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,41 @@
+  register: system_nameservers
+
+- name: check system search domains
+  shell: grep "^search" /etc/resolv.conf | sed 's/search\s*//'
```
ditto

codablock replied:
done

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,41 @@
+
+- name: add system nameservers to docker options
+  set_fact:
+    docker_dns_servers: "{{ docker_dns_servers + [item] }}"
```
union and uniq please

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,41 @@
+
+- name: add system search domains to docker options
+  set_fact:
+    docker_dns_search_domains: "{{ docker_dns_search_domains + [item] }}"
```
union and uniq please

bogdando commented on roles/etcd/tasks/main.yml:
```diff
@@ -15,6 +15,9 @@
   when: is_etcd_master
 
+# Reload systemd before starting service
+- meta: flush_handlers
```
is this really related to the patch?

codablock replied:
It's in a separate commit. When using reset.yml and then cluster.yml afterwards, it for some reason fails in the next task as it can not find the etcd service.

bogdando commented on roles/kubernetes-apps/ansible/tasks/main.yaml:
```diff
@@ -5,7 +5,7 @@
     - {file: kubedns-svc.yml, type: svc}
   register: manifests
-  when: inventory_hostname == groups['kube-master'][0]
+  when: dns_mode != 'none' and inventory_hostname == groups['kube-master'][0]
```
could we merge the `dns_mode='none'` with `dns_setup: false` some how ? this looks a bit confusing

codablock replied:
Hmm, what do you mean with merge?

mattymo replied:
dns_setup variable is now removed completely

codablock commented:
Just committed the changes. I'm still testing out if all combinations work as expected.

mattymo commented on roles/reset/tasks/main.yml:
```diff
@@ -42,6 +42,40 @@
     - /etc/cni
     - /etc/nginx
     - /etc/dnsmasq.d
+    - /etc/dnsmasq.conf
```
I typically dislike unrelated changes attached to a PR, but please avoid doing this in the future.

bogdando commented on roles/reset/tasks/main.yml:
```diff
@@ -42,6 +42,40 @@
+      networking
+      {%- endif %}
+    state: restarted
+  when: ansible_os_family != "CoreOS"
```
we  also don't restart network for canal and calico, there is an unresolved bug

codablock replied:
At this point, there shouldn't be anything alive from canal/calico ;)

bogdando replied:
makes sense :)

bogdando commented:
Please document the changes to the docs/dns-stack.md. Describe how removed dns_setup and skip_dnsmasq map to the new `dns_mode` and `resolvconf_mode` and `hostnet_dns_mode`

codablock commented:
I'll document it after I have retested everything.

codablock commented:
As asked by @bogdando, here is an updated table of supported modes:

dns_mode           |resolvconf_mode|what it does
----------------------|----------------------|-----------------
dnsmasq_kubedns (default) | docker_dns (default) | hosts' unconfigured, docker defines host containers' DNS, dnsmasq is prim resovler
kubedns | docker_dns (default) | hosts' unconfigured, docker defines host containers' DNS, skydns is prim 
dnsmasq_kubedns (default) | host_resolvconf | hosts' resolv.conf configured, docker doesn't, dnsmasq is prim resovler for pods
kubedns | host_resolvconf | hosts' resolv.conf configured, docker doesn't, skydns is prim resovler
none            | *ignored*    | DNS stack is not configured nor for hosts, not docker, nor kubelet and there is no kubedns rc or dnsmasq ds as well

bogdando commented:
Perfect now, thanks!

bogdando commented:
Rerun CI. Please squash when done.

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,57 @@
+- name: add system search domains to docker options
+  set_fact:
+    docker_dns_search_domains: "{{ docker_dns_search_domains + [item] }}"
+  with_items: "{{ system_search_domains.stdout_lines|default([]) }}"
```
where gonna be one line, you should use split('\s')

codablock replied:
Fixed (with .split(' '))

bogdando commented on roles/docker/tasks/set_facts_dns.yml:
```diff
@@ -0,0 +1,57 @@
+
+- name: check number of search domains
+  fail: msg="Too many search domains"
+  when: docker_dns_search_domains|length > 6
```
nit: and total str len <=254

codablock replied:
I've just added this check

codablock commented:
I've just updated the documentation to match the new modes. I hope I covered everything.
I removed the dns.jpeg figure as I'm not able to update the figure. @bogdando I'm not able to update this figure as I don't know how you've created it. Should we actually keep it up to date? To be honest, it was pretty hard to understand the DNS stack from the figure, the documentation and code was actually of much more help.

codablock commented:
CI has kind of passed now. I had to slightly change .gitlab-ci.yml to use the classic host_resolvconf mode for the current CoreOS stable builds as those don't have docker 1.12 yet. This can be removed when CoreOS beta is promoted to stable.

To make sure this works as expected I also manually started the coreos-alpha-weave-ha build which uses the new docker_dns mode. Unfortunately, it turned out that this build is broken on master already. But it looks like docker dns related stuff worked as expected.

I'd say this PR is ready to be merged finally.

bogdando commented:
ci check this

bogdando commented:
Thank you @codablock for a great work done. I liked the docs update and I agree the drawing may be omitted. Let's keep it as an exercise for those who would like to do some reverse engineering.

codablock commented:
Please do not merge. I got strange behavior right now and need to figure out what's wrong

bogdando commented:
@codablock please also rebase for the rkt support. Do you think we should prohibit to use any of the `kubelet_deployment_type: rkt`, `etcd_deployment_type: rkt`  with the `resolvconf_mode=docker_dns`?

codablock commented:
I'd expect that rkt would have no influence on the rkt mode, as it does only deploy kubelet and etcd with rkt and these do not require working cluster DNS. This would be a completely different story if kubernetes would be deployed in the form of rktnetes (where kubelet actually uses rkt to spawn PODs).

Or am I wrong here and the rkt mode already switches kubelet into rktnetes mode?

codablock commented:
ci check this

codablock commented:
ci test this

codablock commented:
Ok, the problem I found is fixed now. PODs were not able to lookup non-cluster names anymore on Azure because dnsmasq was forwarding these queries to 8.8.8.8 (default_resolver). This did not occur on GCE/CI as there is some logic in kargo to use the GCE nameserver as default_resolver.

I added a commit on top that changes dnsmasq to not use the default_resolver when the resolvconf_mode is not host_resolvconf. It will also only set no-resolv option in dnsmasq for the host_resolvconf case.

As I understand it, no-resolv was needed to avoid a loop in lookup as in the past the resolv.conf inside the dnsmasq container was using the hosts resolv.conf, which contained dnsmasq itself.
Now that the hosts resolv.conf is not touched with docker_dns, dnsmasq can rely on the hosts resolv.conf for non-cluster name lookup.

@bogdando Can you review the the commit on top? I also rebased on master, so rkt is included now. All tests passed as expected.
